build-logic/performance-testing/src/main/kotlin/gradlebuild/performance/tasks/BuildCommitDistribution.kt

```diff
@@ -69,21 +87,69 @@ abstract class BuildCommitDistribution @Inject internal constructor(
         val rootProjectDir = project.repoRoot().asFile.absolutePath
         val commit = commitBaseline.map { it.substring(it.lastIndexOf('-') + 1) }
         val checkoutDir = RemoteProject.checkout(fsOps, execOps, rootProjectDir, commit.get(), temporaryDir)
+
+        fsOps.delete { delete(commitDistributionHome.get().asFile) }
         tryBuildDistribution(checkoutDir)
         println("Building the commit distribution succeeded, now the baseline is ${commitBaseline.get()}")
     }
 
+    /**
+     * Sometimes, the nightly might be cleaned up before GA comes out. If this happens, we fall back to latest RC version or nightly version.
+     */
     private
-    fun tryBuildDistribution(checkoutDir: File) {
-        fsOps.delete {
-            delete(commitDistributionHome.get().asFile)
+    fun determineClosestReleasedVersion(expectedBaseVersion: GradleVersion): GradleVersion {
+        val releasedVersions = releasedVersionsFile.asFile.get().reader().use {
+            Gson().fromJson(it, ReleasedVersions::class.java)
+        }
+        if (releasedVersions.finalReleases.any { it.gradleVersion() == expectedBaseVersion }) {
+            return expectedBaseVersion
+        } else if (releasedVersions.latestRc.gradleVersion().baseVersion == expectedBaseVersion) {
+            return releasedVersions.latestRc.gradleVersion()
+        } else if (releasedVersions.latestReleaseSnapshot.gradleVersion().baseVersion == expectedBaseVersion) {
+            return releasedVersions.latestReleaseSnapshot.gradleVersion()
+        } else {
+            throw IllegalStateException("Expected version: $expectedBaseVersion but can't find it")
         }
+    }
+
+    private
+    fun runDistributionBuild(checkoutDir: File, os: OutputStream) {
         execOps.exec {
             commandLine(*getBuildCommands())
             workingDir = checkoutDir
+            standardOutput = os
+            errorOutput = os
+        }
+    }
+
+    private
+    fun tryBuildDistribution(checkoutDir: File) {
+        val output = ByteArrayOutputStream()
+        try {
+            runDistributionBuild(checkoutDir, output)
+        } catch (e: Exception) {
+            val outputString = output.toByteArray().decodeToString()
+            if (failedBecauseOldWrapperMissing(outputString)) {
+                val expectedWrapperVersion = oldWrapperMissingErrorRegex.find(outputString)!!.groups[1]!!.value
+                val closestReleasedVersion = determineClosestReleasedVersion(GradleVersion.version(expectedWrapperVersion)).version
+                val wrapperPropertiesFile = checkoutDir.resolve("gradle/wrapper/gradle-wrapper.properties")
+                val wrapperProperties = Properties().apply {
+                    load(wrapperPropertiesFile.inputStream())
+                    this["distributionUrl"] = "https://services.gradle.org/distributions/gradle-$closestReleasedVersion-bin.zip"
```

Comment:
If we have a snapshot version, then the url needs to be `https://services.gradle.org/distributions-snapshots/...`.

Reply:
Done.